File: Python/n-queen.py
import threading
import time
from datetime import datetime
from typing import Counter

global N
N = 8
count = 0
 
def printSolution(board):
    for i in range(N):
        for j in range(N):
            print (board[i][j], end= " "),
            if j == (N-1):
                print()
 
def isSafe(board, row, col):
 
    # Kontrola na levo
    for i in range(col):
        if board[row][i] == 1:
            return False
 
    # Kontrola horní levá diagonála
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
 
    # Kontrola dolní levá diagonála
    for i, j in zip(range(row, N, 1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
 
    return True
 
def solveNQUtil(board, col):
    if col >= N:
        printSolution(board)
        print("----------")
        globals()['count'] = count + 1
 
    for i in range(N):
 
        if isSafe(board, i, col):
            board[i][col] = 1

            if solveNQUtil(board, col + 1) == True:
                return True

            board[i][col] = 0
 
    return False
 
def solve(row, col):
    board = [['0' for x in range(N)] for y in range(N)]
    if row != -1:
        board[row][0] = 1
 
    if solveNQUtil(board, col) == False:
        print("Reseni neexistuje")
        print("Pocet reseni : " + str(count))
        return False
    
 

def solveParallel():
    threads = []
    for i in range(N):    
        t = threading.Thread(target=solve, args=(i, 1))
        t.start()
        t.join()
        threads.append(t)
 

#solve(-1,0)
#Sériové řešení pro na sobě nezávislý algoritmus
start = datetime.now()
for i in range(N):
    solve(i,1)
end = datetime.now()
print("čas měření : " + str(end-start))

count = 0

# Paralelní řešení pro na sobě nezávislý algoritmus
start = datetime.now()
solveParallel()
end = datetime.now()
print("čas měření : " + str(end-start))

File: Java/NQueen.java
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class NQueen
{
    public static int N = 8;
    public static int count;

    public static void print(int[][] board)
    {
        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < N; j++)
                System.out.print(" " + board[i][j] + " ");
            System.out.println("");
        }
        System.out.println("---------------");
    }

    public static boolean isSafe(int[][] board, int row, int col)
    {
        int i, j;

        // Kontrola na levo
        for (i = 0; i < col; i++)
            if (board[row][i] == 1)
                return false;

        // Kontrola horní levá diagonála
        for (i = row, j = col; i >= 0 && j >= 0; i--, j--)
            if (board[i][j] == 1)
                return false;

        // Kontrola dolní levá diagonála
        for (i = row, j = col; j >= 0 && i < N; i++, j--)
            if (board[i][j] == 1)
                return false;

        return true;
    }

    public static boolean solveNQUtil(int[][] board, int col)
    {
        if (col >= N)
        {
            print(board);
            count++;
        }
        //return true;

        for (int i = 0; i < N; i++)
        {
            if (isSafe(board, i, col))
            {
                board[i][col] = 1;

                if (solveNQUtil(board, col + 1) == true)
                    return true;

                board[i][col] = 0; // BACKTRACK
            }
        }
        return false;
    }

    public static boolean solve(int row, int col)
    {

        int[][] board = new int[N][N];
        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < N; j++)
            {
                board[i][j] = 0;
            }
        }
        if(row != -1)
        {
            board[row][0] = 1; // Aby byl první sloupec statický, tzn. vyzkouší všechna øešení jen pro pevnì danou dámu v prvním sloupci
        }

        if (solveNQUtil(board, col) == false)
        {
        	System.out.println("Reseni neexistuje");
        	System.out.println("Pocet reseni : " + count);
            return false;
        }

       //printSolution(board);
        return true;
    }

    public static boolean solveParallel()
    {
        List<Thread> tasks = new ArrayList<>();
        for (int i = 0; i < N; i++)
        {
        	final int row = i;
            Thread t = new Thread(() -> solve(row, 1));
            t.start();
            try {
				t.join();
			} catch (Exception ex) {
				System.out.print("chyba");
			}
            tasks.add(t);
        }
        return true;
    }
}